Connections management refactor

ricardogsilva opened the pull request:
This PR refactors the `connections_management` branch (which is currently being PR'd for inclusion into main) in order to fix some of the outstanding issues I found during my latest review. Notable changes:

- introduce the `conf.settings_manager` object, which is now responsible for interacting with `QgsSettings` on behalf of the plugin
- rename some of the ui files and some attributes of UI classes
- properly restore a connection's authcfg - there was a bug in the previous implementation whereby the authcfg was being stored with the wrong settings key

To the best of my knowledge everything is now working properly (added some basic tests, to be expanded)

@Samweli please study this PR and verify that it works. After that, merge it into your branch and then request a review from me on the main PR.

Samweli committed:
adding connections management

Commit:
Merge remote-tracking branch 'Samweli/connections_management' into connections_management

# Conflicts:
#	src/qgis_geonode/gui/connection_dialog.py
#	src/qgis_geonode/gui/geonode_source_select_provider.py
#	src/qgis_geonode/ui/qgis_geonode_connection.ui

Commit:
Start refactoring connections management

Commit:
finish refactor of connection management

Commit:
format code with black

Samweli commented:
> @Samweli please study this PR and verify that it works.

noted @ricardogsilva

Samweli commented on src/qgis_geonode/qgisgeonode/conf.py:
```diff
@@ -0,0 +1,85 @@
+    def save_connection_settings(self, name, **additional_settings):
+        with qgis_settings(self._get_connection_settings_base(name)) as settings:
+            for name, value in additional_settings.items():
+                settings.setValue(name, value)
```
when editing a connection, we might need to delete the past connection settings here. I have tested this on QGIS, another connection is created when editing a connection name.

ricardogsilva replied:
I solved this by adding an additional `id` to each `ConnectionSettings` instance. This `id` field is not shown to the user and remains stable

Samweli commented on src/qgis_geonode/gui/geonode_source_select_provider.py:
```diff
@@ -61,99 +40,65 @@ def ordering(self):
-        self.create_connections_list()
+        settings_manager.current_connection_changed.connect(
+            self.update_connections_combobox
+        )
```
we might need to update the combo box connection here, currently connections that were added, don't show up when the widget is loaded for the first time.

Commit:
finish connections management refactor

- each connection is now represented as an instance of `ConnectionSettings` instead of a dict
- add an immutable (i.e. hidden from the user) `id` property to each connection in order to avoid using the name as the connection identifier
- add a `test connection` button in the connection dialog to test if connection parameters work
- add initial refresh of existing connections in order to make sure they are shown on the GUI
- set the current connection in case it does not exist yet

Samweli commented on src/qgis_geonode/gui/geonode_source_select_provider.py:
```diff
@@ -54,51 +54,67 @@ def __init__(self, parent, fl, widgetMode):
+        if current_connection is None:
+            existing_connections = connections_manager.list_connections()
+            if len(existing_connections) > 0:
+                current_connection = existing_connections[0]
```
shouldn't we default to the last added connection instead of the first one?

ricardogsilva replied:
Since they are being sorted by name (check the implementation of `conf.ConnectionManager.list_connections` this is a fairly arbitrary choice. Picking the first one or the last does not imply it is more recent.

Or are you suggesting we add a new temporal field to each `ConnectionSettings` instance to track when it was created and then sort by that? IMO it sounds interesting, but I don't think it is relevant to implement such a feature at the moment

Samweli replied:
> Or are you suggesting we add a new temporal field to each `ConnectionSettings` instance to track when it was created and then sort by that?

no, I expected the connection list have the recent connection at the last index.

> IMO it sounds interesting, but I don't think it is relevant to implement such a feature at the moment

sure

Samweli commented on src/qgis_geonode/qgisgeonode/conf.py:
```diff
@@ -0,0 +1,141 @@
+    @classmethod
+    def from_qgs_settings(cls, connection_identifier: str, settings: QgsSettings):
+        return cls(
+            id=uuid.UUID(connection_identifier),
```
we have to check for `ValueError('badly formed hexadecimal UUID string')` here

ricardogsilva replied:
I don't seem to have that error.

can you please try to empty your relevant `QgsSettings` and try again? Try this in the QGIS Python console:

```python
from qgis_geonode.qgisgeonode.conf import connections_manager
connections_manager.delete_all_connections()
connections_manager.clear_current_connection()
```

then try again and see if you get that error. If you do I will fix it, but I suspect it is not going to happen anymore

Samweli replied:
ok, let me try it.

Samweli replied:
yes, the error disappeared!

Samweli commented on src/qgis_geonode/gui/connection_dialog.py:
```diff
@@ -1,84 +1,91 @@
+            auth_config=self.authcfg_acs.configId(),
+        )
+
+    def test_connection(self):
```
we should disable the "Test Connection" button and enable it after finishing testing connection.

ricardogsilva replied:
:+1: I can add this in

Commit:
Toggle connection dialog buttons state while test is running
